resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsEventHandler.scala

```diff
@@ -0,0 +1,229 @@
+  // take precedence.
+  private val runningExecutors = mutable.Set.empty[Long]
+
+  private var eventProcessorFuture: Future[_] = _
+
+  def start(applicationId: String, schedulerBackend: KubernetesClusterSchedulerBackend): Unit = {
+    require(eventProcessorFuture == null, "Cannot start event processing twice.")
+    logInfo(s"Starting Kubernetes executor pods event handler for application with" +
+      s" id $applicationId.")
+    val eventProcessor = new Runnable {
+      override def run(): Unit = {
+        Utils.tryLogNonFatalError {
+          processEvents(applicationId, schedulerBackend)
+        }
+      }
+    }
+    eventProcessorFuture = eventProcessorExecutor.scheduleWithFixedDelay(
+      eventProcessor, 0L, podAllocationDelay, TimeUnit.MILLISECONDS)
+  }
+
+  def stop(): Unit = {
+    if (eventProcessorFuture != null) {
+      eventProcessorFuture.cancel(true)
+      eventProcessorFuture = null
+    }
+  }
+
+  private def processEvents(
+      applicationId: String, schedulerBackend: KubernetesClusterSchedulerBackend) {
+    val currentEvents = new java.util.ArrayList[Seq[Pod]](eventQueue.size())
+    eventQueue.drainTo(currentEvents)
+    currentEvents.asScala.flatten.foreach { updatedPod =>
+      val execId = updatedPod.getMetadata.getLabels.get(SPARK_EXECUTOR_ID_LABEL).toLong
+      val podPhase = updatedPod.getStatus.getPhase.toLowerCase
+      if (isDeleted(updatedPod)) {
+        removeExecutorFromSpark(schedulerBackend, updatedPod, execId)
+      } else {
+        updatedPod.getStatus.getPhase.toLowerCase match {
+          case "running" =>
+            // If clause is for resililence to out of order operations - executor must be
+            // pending and first reach running. Without this check you may e.g. process a
+            // deletion event followed by some arbitrary modification event - we want the
+            // deletion event to "stick".
+            if (pendingExecutors.contains(execId)) {
+              pendingExecutors.remove(execId)
+              runningExecutors.add(execId)
+            }
+          // TODO (SPARK-24135) - handle more classes of errors
+          case "error" | "failed" | "succeeded" =>
+            // If deletion failed on a previous try, we can try again if resync informs us the pod
+            // is still around.
+            // Delete as best attempt - duplicate deletes will throw an exception but the end state
+            // of getting rid of the pod is what matters.
+            if (!isDeleted(updatedPod)) {
+              Utils.tryLogNonFatalError {
+                kubernetesClient
+                  .pods()
+                  .withName(updatedPod.getMetadata.getName)
+                  .delete()
+              }
+            }
+            removeExecutorFromSpark(schedulerBackend, updatedPod, execId)
+        }
+      }
+    }
+
+    val currentRunningExecutors = runningExecutors.size
```

Comment:
I think you can move the code for creating new executors to its own executor service that can actually run concurrently with the executor service for processing pod events. This way you can use independent schedules for the two. I think this is less confusing and allows pod event processing to run with a higher frequency than pod allocation.

Reply:
I thought about this, and prefer the current way because of the ease of reasoning about synchronization. One of the downsides with the previous approach was the need to synchronize state and reason about thread safety in multiple places. With a separate executor service that reads/writes from the `pendingExecutorIds` and `runningExecutorIds` field, we'd have to reason about how the two executors would share about these data structures with the right thread safety model.

But what do you think, would the concurrency complexity be manageable and worth the upside of having separate schedules?

Reply:
Furthermore I don't think it's just a matter of using thread safe data structures and access to the hash sets, but one also has to think carefully about ordering and atomicity, etc.

Reply:
I think the concurrency complexity is totally manageable. The idiom of Kubernetes controllers is you should not be worrying about ordering of events, but instead should be level-triggered. I don't think it makes much sense to use the pod allocation delay as the interval for both executor pod allocation and executor pod event processing.

Reply:
I like the idea of not having this method be a single large method and not having this class need to deal with everything. I'll try a patch that separates out the executor construction from the event handling, but, once that patch is out please review it carefully to ensure that the concurrency semantics are correct.

Reply:
Had another idea - you can denormalize the state, meaning each event processor module maintains its own copy of the perceived state. We get the modularization and separate scheduling, without having to worry about thread safety. The amount of state being duplicated is quite small. I'll propose the idea and would appreciate feedback once it's in.

Reply:
Sounds good to me.